### PR TITLE
Several minor improvements

### PR DESCRIPTION
**Pull request by sennendoko**
This PR contains several minor improvements:

- General improvements to the wrappers used in the library to use Gaussian process models from GPy. The documentation of the wrapper has been improved as well as the code style. A couple of minor standard algorithmic improvements have been implemented.
- A bug has been corrected in the way the library deals with categorical parameters.
- General improvements in the notebooks of the library.
- Improvements in some unit tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by apaleyes** [comment]
Thanks for the contribution, @sennendoko. We would need to take our time with this one though. To begin with, I would like to ask you to break it down into separate purposeful PRs, if it all possible. Some of the advantages of doing it this way:
- It is much easier to review when all changes in the PR belong to the same scope.
- It keeps the commit history meaningful. When merging this PR, I would need to squash everything together.
- It is easier to have a conversation that others can follow and chime in, if necessary. Conversations in this PR are likely to be about many changes, and it would be hard to follow any single one of them.

At the very least I would ask to separate out changes about categorical variables handling, and changes to local penalization acquisition, and these appear to be more severe and certainly warrant separate discussion, which I'd much rather have in separate threads.



Regarding other changes in this PR, I will refrain from reviewing them just yet, until we decide on the question above.

**Comment by sennendoko** [comment]
OK - will split into separate PRs, and close this one.

**Comment by apaleyes** [comment]
Thank you, that will make our work on these new PRs much more efficient!
